test/e2e/virtual/workspaces/virtual_workspace_test.go

```diff
@@ -155,69 +171,49 @@ func TestWorkspacesVirtualWorkspaces(t *testing.T) {
 			},
 			work: func(ctx context.Context, t *testing.T, server runningServer) {
 				vwUser1Client := server.virtualWorkspaceClients[0]
-				_, err := server.kubeClient.CoreV1().Namespaces().Create(ctx, &v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "default"}}, metav1.CreateOptions{})
+				_, err := server.orgKubeClient.CoreV1().Namespaces().Create(ctx, &v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "default"}}, metav1.CreateOptions{})
 				require.NoError(t, err, "failed to create namespace")
 
 				kcpServerKubeconfig, err := server.RunningServer.RawConfig()
 				require.NoError(t, err, "failed to get KCP Kubeconfig")
 
-				kubeconfigContent, err := clientcmd.Write(kcpServerKubeconfig)
-				require.NoError(t, err, "failed to get KCP Kubeconfig content: %v")
-
-				_, err = server.kubeClient.CoreV1().Secrets("default").Create(ctx, &v1.Secret{
-					ObjectMeta: metav1.ObjectMeta{
-						Name:      "kubeconfig",
-						Namespace: "default",
-					},
-					StringData: map[string]string{
-						"kubeconfig": string(kubeconfigContent),
-					},
-				}, metav1.CreateOptions{})
-				require.NoError(t, err, "failed to create workspace shard secret")
-
-				_, err = server.kcpClient.TenancyV1alpha1().WorkspaceShards().Create(ctx, &tenancyv1alpha1.WorkspaceShard{
-					ObjectMeta: metav1.ObjectMeta{
-						Name: "boston",
-					},
-					Spec: tenancyv1alpha1.WorkspaceShardSpec{
-						Credentials: v1.SecretReference{
-							Name:      "kubeconfig",
-							Namespace: "default",
-						},
-					},
-				}, metav1.CreateOptions{})
-				require.NoError(t, err, "failed to create workspace shard")
-
-				server.Artifact(t, func() (runtime.Object, error) {
-					return server.kcpClient.TenancyV1alpha1().WorkspaceShards().Get(ctx, "boston", metav1.GetOptions{})
-				})
-
-				workspace1, err := vwUser1Client.TenancyV1alpha1().Workspaces().Create(ctx, testData.workspace1.DeepCopy(), metav1.CreateOptions{})
+				workspace1, err := vwUser1Client.TenancyV1beta1().Workspaces().Create(ctx, testData.workspace1.DeepCopy(), metav1.CreateOptions{})
 				require.NoError(t, err, "failed to create workspace1")
 
 				server.Artifact(t, func() (runtime.Object, error) {
-					return server.kcpClient.TenancyV1alpha1().Workspaces().Get(ctx, testData.workspace1.Name, metav1.GetOptions{})
+					return server.orgKcpClient.TenancyV1alpha1().ClusterWorkspaces().Get(ctx, testData.workspace1.Name, metav1.GetOptions{})
 				})
 
 				workspaceURL := ""
-				err = server.kcpExpect(workspace1, func(w *tenancyv1alpha1.Workspace) error {
-					if !conditions.IsTrue(w, tenancyv1alpha1.WorkspaceURLValid) {
-						return fmt.Errorf("Workspace %s is not valid: %s", w.Name, conditions.GetMessage(w, tenancyv1alpha1.WorkspaceURLValid))
+				var lastErr error
+				err = wait.PollImmediate(time.Millisecond*100, wait.ForeverTestTimeout, func() (done bool, err error) {
+					defer func() {
```

Comment:
(No action required) Rather than jumping through hoops to save the last err, maybe log the error? Also, maybe don't return err from Get to allow retry on potentially transient failures.

Reply:
for followup: a require.Eventually variant with a way to return errors.

Reply:
Is it not sufficient to log an error? Returning only the last error seems less-than-ideal where multiple errors could occur during the polling interval.